### PR TITLE
broker: add cleanup phase

### PR DESCRIPTION
**Pull request by garlick**
This PR moves the following stanza out of the rc3 script, as proposed in #2909
```sh
flux queue stop --quiet
flux job cancelall --user=all --quiet -f --states RUN
flux queue idle --quiet
```
A new mechanism for pushing cleanup commands onto a list that runs after the initial program finishes is now available.  The cleanup command list has the following semantics:
* Cleanup runs whether the initial program succeeds or fails
* Cleanup does not run when rc1 fails (as the initial program does not run  in that case)
* A cleanup failure causes the broker to fail (exit with exit code of the failing cleanup command)
* If one command in the list fails, cleanup commands continue to execute
* If multiple cleanup commands fail, the exit code of the first failure is propagated to broker exit code

The cleanup list can be prepended by running `flux admin cleanup-push command ...`.  The default `rc1` script now pushes the above commands onto the cleanup list.

If a workflow or batch script wants to use this facility as instance owner, they are welcome to push additional commands.

Internally, some broker refactoring was needed to make this possible.  The "runlevel" code was replaced with a generic "runat" class and a state machine.  The same code that implements the "cleanup" command list also runs rc1, 2 and 3.  I think the "state machine" approach, as opposed to the hardwired runlevels,  will make implementing the distributed rc1/rc3 proposed in #2614 less hacky,.

### CODE REVIEW

**Comment by garlick** [comment]
Just pushed some test improvements, and got rid of a hueristic I was using to shorten logged command lines (did more harm than good IMHO).  This is probably ready for a review.

**Comment by garlick** [comment]
Just forced a push to
* rebase on current master to drop @grondo's mypy fix that was merged
* fix the aforementioned unintended change
* improve commit messages a bit

**Comment by garlick** [comment]
restarted centos 7 buider that had a broker timeout here
```
# passed all 42 test(s)
flux-start: 0 (pid 20070) Killed
1..42
ERROR: t1005-kvs-security.t - exited with status 137 (terminated by signal 9?)
```

File: etc/rc1 [hunk 1]
```diff
@@ -50,3 +50,8 @@ for rcdir in $all_dirs; do
     done
 done
 shopt -u nullglob
+
+# N.B. push cleanup commands in the reverse order of execution
+flux admin cleanup-push "flux queue idle --quiet"
+flux admin cleanup-push "flux job cancelall --user=all --quiet -f --states RUN"
+flux admin cleanup-push "flux queue stop --quiet"
```
Comment (grondo) [inline etc/rc1:57]: Just a thought - would it be more straightforward to have an append version of `flux admin cleanup`? This is the only current use case of `flux admin cleanup-push", and it requires an "N.B." comment, so it gave me pause.

Comment (grondo) [reply etc/rc1:57]: Or, maybe `flux admin cleanup-push` could take a script on stdin and could be called like:
```
flux admin cleanup-push <<EOF
flux queue stop --quiet
flux job cancelall --user=all --quiet -f --states RUN
flux queue idle --quiet
EOF
```

Comment (garlick) [reply etc/rc1:57]: I like the stdin idea!

**Comment by garlick** [comment]
I pushed a few commits to implement the `flux admin cleanup-push <<-EOT` idea.  Let me know if that seems OK and I'll squash.

**Comment by garlick** [comment]
Squashed.  OK to set MWP?

**Comment by grondo** [comment]
Sounds good to me!

**Pull request merged**